tests/Unit/controller/FrontController/ProductControllerTest.php

```diff
@@ -0,0 +1,310 @@
+<?php
+/**
+ * 2007-2017 PrestaShop
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to http://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2017 PrestaShop SA
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShop\PrestaShop\Tests\Unit\Controller\FrontController;
+
+use PrestaShop\PrestaShop\Adapter\Product\PriceFormatter;
+
+class ProductControllerTest extends \PrestaShop\PrestaShop\Tests\TestCase\IntegrationTestCase
+{
+    private $controller;
+    private $contextBackup;
+
+    public function setUp()
+    {
+        parent::setUp();
+        $this->contextBackup = \Context::getContext();
+        $context             = clone($this->contextBackup);
+        \Context::setInstanceForTesting($context);
+        $context->shop       = new \Shop((int) \Configuration::get('PS_SHOP_DEFAULT'));
+        $context->language   = new \Language;
+        $protocol_link       = (\Tools::usingSecureMode() && \Configuration::get('PS_SSL_ENABLED'))
+            ? 'https://' : 'http://';
+        $protocol_content    = (\Tools::usingSecureMode() && \Configuration::get('PS_SSL_ENABLED'))
+            ? 'https://' : 'http://';
+        $context->link       = new \Link($protocol_link, $protocol_content);
+        $context->smarty     = new \Smarty();
+        $this->controller = new \ProductControllerCore();
+    }
+
+    public function tearDown()
+    {
+        parent::tearDown();
+
+        \Context::setInstanceForTesting($this->contextBackup);
+    }
+
+    /**
+     * Call protected/private method of a class.
+     *
+     * @param object $object     Instantiated object that we will run method on.
+     * @param string $methodName Method name to call
+     * @param array  $parameters Array of parameters to pass into method.
+     *
+     * @return mixed Method return.
+     */
+    public function invokeMethod($object, $methodName, array $parameters = [])
```

Comment:
Before someone else says it... we only use the long array syntax `array()` for some reason that I keep forgetting.

Reply:
done

Reply:
You forgot it here 😛